### PR TITLE
issue #520: provide option to customize stock extensions

### PR DESCRIPTION
**Pull request by pepperbob**
This is a draft for #520 to allow for a customization of the default extensions.

- refactored out the extension loading code from the `PebbleEngine.Builder` into the `ExtensionRegistryFactory` 
- provide a `ExtensionCustomizer` which basically allows to access extensions before loading

### CODE REVIEW

**Comment by pepperbob** [comment]
So I came up with "ExtensionCustomizer" to allow filtering out or change whatever functionality is provided by an extension. It currently only targets the baked in extensions - not sure if it would make much sense on the user provided extensions anyway. 

On the one hand it looks like it leaves the configuration API completely intact and doesn't degrade things like `allowOverrideCoreOperators` to an implementation detail.

On the other hand I'm not too happy with the current version as now other implementation details, e.g. how core functionality is split over extension classes (core, i18n, escaper) kind of leaks out as you have to know the extension class to apply the customizer.

Of course the whole target class thing can be dropped and there could be just one single customizer that is wrapped around every extension before it is loaded ... while thinking about it - i think this is would would provide the cleanest API and won't break things for users if stuff is re-organized internally at some point.. I'll rework it.

Feedback is much appreciated!

File: pebble/src/test/java/com/mitchellbosecke/pebble/extension/ExtensionCustomizerTest.java [hunk 1]
```diff
@@ -37,15 +39,16 @@ void upperFilterCannotBeUsed() throws IOException {
...
 
-    public RemoveUpper(Extension core) {
+    public RemoveUpperCustomizer(Extension core) {
       super(core);
     }
 
     @Override
     public Map<String, Filter> getFilters() {
-      Map<String, Filter> filters = new HashMap<>(super.getFilters());
+      Map<String, Filter> filters = Optional.ofNullable(super.getFilters()).map(HashMap::new)
```
Comment (pepperbob) [inline pebble/src/test/java/com/mitchellbosecke/pebble/extension/ExtensionCustomizerTest.java:50]: just curious: is there a reason why null is acceptable over an empty list if filters and others are not provided?

**Comment by ebussieres** [comment]
Can you provide some documentation for this new functionality ?

**Comment by pepperbob** [comment]
@ebussieres Ok, I added further JavaDoc - is that okay so far?

**Comment by ebussieres** [comment]
I mean add some documentation for pebble website. Doc is here : https://github.com/PebbleTemplates/pebble/tree/master/docs/src/orchid/resources

**Comment by pepperbob** [comment]
Oh sure, I'll check this out!

File: docs/src/orchid/resources/wiki/guide/customize-defaults.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+In these cases it might be advised to consider stripping-down Pebbles' built-in functionality that may otherwise introduce security-concers regarding the integrity and stability of your application.
+
+### Opt-Out using ExtensionCustomizer
+
+The `ExtensionCustomizer` base class can be used to gain access to the default functionality before it is loaded into Pebbles template engine. Overwrite methods to get hold on provided default-functionality and modify whatever should be available for the template engine.
+
+The following example removes the `ForTokenParser`, i.e. the ability to parse `{% for %}{{ ... }}{% endfor %}` constructs:
+
+```java
+class ExampleOptOuts extends ExntensionCustomizer {
```
Comment (ebussieres) [inline docs/src/orchid/resources/wiki/guide/customize-defaults.md:15]: there's a typo in `ExntensionCustomizer`

Comment (pepperbob) [reply docs/src/orchid/resources/wiki/guide/customize-defaults.md]: :facepalm: fixed it :)

**Pull request merged**